xds/internal/balancer/gracefulswitch/gracefulswitchbalancer.go

```diff
@@ -0,0 +1,374 @@
+	// balancer neither current or pending) can get interspliced here before
+	// UpdateAddresses() - however, doesn't seem to be a problem with regard to
+	// correctness.
+	gsb.cc.UpdateAddresses(sc, addrs) // Can this call back inline?
+}
+
+func (gsb *gracefulSwitchBalancer) SwitchTo(builder balancer.Builder) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	ccw := &clientConnWrapper{
+		ClientConn: gsb.cc,
+		gsb:        gsb,
+	}
+	newBalancer := builder.Build(ccw, gsb.bOpts)
+	if newBalancer == nil {
+		// Can this even ever happen?
+		return balancer.ErrBadResolverState
+	}
+	ccw.bal = newBalancer
+	if gsb.balancerCurrent == nil {
+		gsb.balancerCurrent = newBalancer
+		gsb.mu.Unlock()
+		return nil
+	}
+	var balToClose balancer.Balancer
+	gsb.pendingState = balancer.State{
+		ConnectivityState: connectivity.Connecting,
+		Picker:            base.NewErrPicker(balancer.ErrNoSubConnAvailable),
+	}
+	// Clean up resources here that are from a previous pending lb.
+	for sc, sb := range gsb.scToSubBalancer {
+		if sb == gsb.balancerPending {
+			gsb.cc.RemoveSubConn(sc)
+			delete(gsb.scToSubBalancer, sc)
+		}
+	}
+	// Need to test this by verifying somehow.
+	balToClose = gsb.balancerPending
+	gsb.balancerPending = newBalancer
+	gsb.mu.Unlock()
+	if balToClose != nil {
+		balToClose.Close()
+	}
+	return nil
+}
+
+// UpdateClientConnState() simply forwards the update to one of it's children.
+func (gsb *gracefulSwitchBalancer) UpdateClientConnState(state balancer.ClientConnState) error {
+	if gsb.closed.HasFired() {
+		// logger?
+		return errBalancerClosed
+	}
+	gsb.mu.Lock()
+	balToUpdate := gsb.balancerPending
+	if balToUpdate == nil {
+		balToUpdate = gsb.balancerCurrent
+	}
+	gsb.mu.Unlock()
+	balToUpdate.UpdateClientConnState(state)
+	return nil
+}
+
+func (gsb *gracefulSwitchBalancer) ResolverError(err error) {
+	if gsb.closed.HasFired() {
+		// Logger?
+		return
+	}
+	gsb.mu.Lock()
+	currentBalancerToUpdate := gsb.balancerCurrent
+	pendingBalancerToUpdate := gsb.balancerPending
+	gsb.mu.Unlock()
+	// The update will be forwarded to the pending balancer only if there is a
+	// pending LB present, as that is the most recent LB Config prepared by the
+	// resolver, and thus is a separate concern from the current.
+	if currentBalancerToUpdate != nil && pendingBalancerToUpdate == nil {
+		currentBalancerToUpdate.ResolverError(err)
+	}
+	if pendingBalancerToUpdate != nil {
+		pendingBalancerToUpdate.ResolverError(err)
+	}
+}
+
+func (gsb *gracefulSwitchBalancer) UpdateSubConnState(sc balancer.SubConn, state balancer.SubConnState) {
+	if gsb.closed.HasFired() {
+		// Logger
+		return
+	}
+	gsb.mu.Lock()
+	bal, ok := gsb.scToSubBalancer[sc]
+	if !ok {
+		gsb.mu.Unlock()
+		return
+	}
+	if state.ConnectivityState == connectivity.Shutdown {
+		delete(gsb.scToSubBalancer, sc)
+	}
+	gsb.mu.Unlock()
+	// After giving up the mutex, the current balancer may be Closed() due to a
```

Comment:
It is impossible for UpdateClientConnState or ResolverError to induce this race condition. The reason is that a balnacer.Close() call on the current can only happen when there is both a pending and a current balancer. However, since in the situation where there is both a current and pending balancer in UpdateClientConnState() or ResolverError(), the call will always be forwarded to the pending balancer. Thus, it is impossible for the race to happen between balancer Close() and Update(), and thus UpdateClientConnState() and ResolverError() will never break the balancer API guarantee of no Update() after Close().

Reply:
See https://github.com/grpc/grpc-go/pull/5117/files#diff-7467ada1ef249617fb2ad7dfe22c6fbbfe54c05cfe1c9de992373c16630dff23R876 for the test/some explanation.